util/file_reader_writer.cc

```diff
@@ -37,7 +60,24 @@ Status RandomAccessFileReader::Read(uint64_t offset, size_t n, Slice* result,
     StopWatch sw(env_, stats_, hist_type_,
                  (stats_ != nullptr) ? &elapsed : nullptr);
     IOSTATS_TIMER_GUARD(read_nanos);
-    s = file_->Read(offset, n, result, scratch);
+    if (UseDirectIO()) {
+      size_t alignment = file_->GetRequiredBufferAlignment();
+      size_t aligned_offset = TruncateToPageBoundary(alignment, offset);
+      size_t offset_advance = offset - aligned_offset;
+      size_t size = Roundup(offset + n, alignment) - aligned_offset;
+      AlignedBuffer buf;
+      buf.Alignment(alignment);
+      buf.AllocateNewBuffer(size);
+      Slice tmp;
+      s = file_->Read(aligned_offset, size, &tmp, buf.BufferStart());
+      if (s.ok()) {
+        buf.Size(tmp.size());
+        size_t r = buf.Read(scratch, offset_advance, tmp.size() <= offset_advance ? 0 : std::min(tmp.size() - offset_advance, n));
+        *result = Slice(scratch, r);
+      }
+    } else {
+      s = file_->Read(offset, n, result, scratch);
+    }
```

Comment:
Create a RandomAccessFileReader::ReadDirectIO() and deal with the direct I/O case, and this function will still be short. It makes it easier for this function to be inlined for non-dirct-I/O case, especially the mmap case, which is very performance critical.